File: www/src/lib/api/asns.ts
import { apiFetch } from "@/lib/api/client";
import type {
  EntityPeakStats,
  PlayersSummaryBase,
  PlayersTimeseriesPayload,
} from "@/lib/api/types";
import { fetchList } from "@/lib/api/types";
import type { ServerListItem } from "@/lib/api/servers";

export type AsnsSummary = PlayersSummaryBase & {
  trackedAsns: number;
  trackedServers: number;
};

export type AsnListItem = {
  asn: string;
  asnOrg: string;
  playersOnline: number;
  serverCount: number;
  peaks: EntityPeakStats;
};

export type AsnsListResponse = {
  summary: AsnsSummary;
  asns: AsnListItem[];
};

export type AsnDetailResponse = {
  asn: string;
  asnOrg: string;
  playersOnline: number;
  serverCount: number;
  peaks: EntityPeakStats;
  summary: PlayersSummaryBase & { trackedServers: number };
  servers: ServerListItem[];
};

export type AsnTimeseriesResponse = PlayersTimeseriesPayload & {
  asn: string;
  asnOrg: string;
};

export type AsnDetailSearch = {
  asnOrg?: string;
};

export function asnDetailSearch(asnOrg: string): AsnDetailSearch {
  const trimmed = asnOrg.trim();
  return trimmed.length > 0 ? { asnOrg: trimmed } : {};
}

export function asnDisplayName(asn: Pick<AsnListItem, "asn" | "asnOrg">): string {
  if (asn.asnOrg) {
    return asn.asnOrg;
  }
  if (asn.asn) {
    return asn.asn;
  }
  return "Unknown network";
}

export function getAsns(search?: string) {
  return fetchList<AsnsListResponse>("/asns", search);
}

export function getAsn(asn: string, asnOrg = "") {
  const params = new URLSearchParams();
  if (asnOrg) {
    params.set("asnOrg", asnOrg);
  }
  const query = params.toString();
  const path = query ? `/asns/${encodeURIComponent(asn)}?${query}` : `/asns/${encodeURIComponent(asn)}`;
  return apiFetch<AsnDetailResponse>(path, { credentials: "omit" });
}

export function getAsnTimeseries(
  asn: string,
  asnOrg: string,
  from: number,
  to: number,
) {
  const params = new URLSearchParams({
    asn,
    from: String(from),
    to: String(to),
  });
  if (asnOrg) {
    params.set("asnOrg", asnOrg);
  }
  return apiFetch<AsnTimeseriesResponse>(`/asns/timeseries?${params}`, {
    credentials: "omit",
  });
}

File: crates/mc-api-types/src/response/asns.rs
use serde::Serialize;

use super::servers::{
    EntityPeakStats, PlayersPeakSummary, ServerListItemResponse, ServersSummaryResponse,
};
use super::timeseries::TimeseriesLanes;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AsnsListResponse {
    pub summary: AsnsSummaryResponse,
    pub asns: Vec<AsnListItemResponse>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AsnsSummaryResponse {
    pub total_players: u64,
    pub players_pc: u64,
    pub players_pe: u64,
    pub tracked_asns: u32,
    pub tracked_servers: u32,
    pub peaks: PlayersPeakSummary,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AsnListItemResponse {
    pub asn: String,
    pub asn_org: String,
    pub players_online: u32,
    pub server_count: u32,
    pub peaks: EntityPeakStats,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AsnDetailResponse {
    pub asn: String,
    pub asn_org: String,
    pub players_online: u32,
    pub server_count: u32,
    pub peaks: EntityPeakStats,
    pub summary: ServersSummaryResponse,
    pub servers: Vec<ServerListItemResponse>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AsnTimeseriesResponse {
    pub asn: String,
    pub asn_org: String,
    #[serde(flatten)]
    pub timeseries: TimeseriesLanes,
}
